### ISSUE TITLE
Harden instrumentation-state: make integration telemetry completeness self-verifying

### ISSUE THREAD

**Issue by Pendu** [traceopt-ai/traceml#141]
**Status:** Stage 1 PR is open: #151 (rebased onto current `main`; additive, `359 passed / 2 skipped`). Self-assigned to @Pendu.

## Summary
TraceML installs its instrumentation via **process-global, mutable state** — sentinels like `nn.Module._traceml_forward_patched`, `Tensor._traceml_h2d_patched`, `Optimizer._traceml_opt_hooks_installed`, and the `DataLoader.__iter__` swap — set by **three overlapping paths** (the CLI launcher's `init()`, an explicit `init()`, and a legacy import-time side-effect) and **reset/owned by none**.

Two consequences fall out:
- An integration can be individually correct yet silently emit **incomplete telemetry** (a stream is dark, and nothing says so).
- Tests can pass **for the wrong reasons** (global state leaks across them).

This surfaced during the callback-based HF integration (#88 / #135): the callback path didn't *guarantee* DataLoader-fetch timing — it only worked *incidentally* because importing the integration triggers the legacy auto-init side-effect. The gap was caught by review, not by the system. We'd like completeness to be a **structural, self-verifying property**.

## Evidence (reproducible)
**A — import-time install is incidental, not guaranteed.** Pristine process, no `init()` call, no prior tests — importing the integration alone initializes the process and patches the DataLoader:
```bash
PYTHONPATH=src python3 -c "
import torch.utils.data as tud; before = tud.DataLoader.__iter__
from traceml_ai.integrations.huggingface import TraceMLTrainerCallback
from traceml_ai.sdk.initial import is_initialized
print('is_initialized:', is_initialized())                              # True
print('DataLoader patched by import:', before is not tud.DataLoader.__iter__)  # True
"
```
**B — global state never reset → order-dependent test failure** (reproduces on `main`):
```bash
PYTHONPATH=src python3 -m pytest \
  tests/integrations/test_hf_trainer.py \
  "tests/sdk/test_init_and_wrappers.py::test_wrap_optimizer_preserves_identity_and_times_step" -q
# -> wrap_optimizer test FAILS: "global optimizer step hooks are already installed"
```
Two symptoms, **one root cause** (ungoverned global instrumentation state): (A) a *missing-install* masked by an incidental side-effect; (B) a *never-uninstall* that poisons sibling tests.

## Why feature-level tests don't catch this
The integration is pure *mechanism* and is correct; *which streams are actually live* is *policy* in separate global state. The two are decoupled, so a feature-scoped test can't observe a dark stream — the failure is an **absence**, invisible in a diff.

## Proposed approach (staged)
**Stage 1 — additive, zero behavior change (first PR):**
1. **Cross-integration conformance gate** — each integration *declares* the telemetry streams it owes; a parametrized test asserts each is emitted via the documented `init()` path. (Wires HF now; Lightning when it's a test dependency; Ray excluded — orchestrator-level.)
2. **DataLoader-fetch completeness guard** for the HF path (asserts the #88-class stream actually lands).
3. **Fail-loud capability assertion** — when an integration is active but the required patches aren't configured, warn (never raise) which streams will be dark.
4. **Optimizer-hook reset** — public reset + autouse teardown clearing the global optimizer-hook handles/flag between tests → suite order-independent (fixes symptom B).

**Stage 2 — back-compat-sensitive (separate follow-up PR):** eliminate the legacy import-time side-effect and make explicit `init()` the single install path (with a deprecation window); this unlocks a deterministic negative-control conformance test (the regression that would have caught #88). Also extend the contract to `step_memory` + GPU `h2d`, and wire the capability assertion into the Lightning integration.

**Update (#151 rebased onto `main` @ `95483c7`):** upstream #146 already removed the legacy import-time auto-init (`sdk/decorators_compat`), so Stage 2's first item is effectively done upstream. Stage 2 now reduces to: the deterministic negative-control conformance test, extending the contract to `step_memory` + GPU `h2d`, and the Lightning conformance harness + capability wiring.

## Acceptance (Stage 1)
- [ ] Parametrized conformance test asserts each declared stream is emitted end-to-end (HF); fails if one is missing.
- [ ] Integration active-but-uninitialized → clear best-effort warning naming the dark streams (no crash).
- [ ] Full suite passes regardless of test order; regression test for the order-dependence.
- [ ] No training-path behavior/overhead change; all existing tests pass.
